en/lessons/libraries/bypass.md

```diff
@@ -0,0 +1,274 @@
+To accomplish this we won't be using `Bypass.expect/2` instead we're going to rely on `Bypass.down/1` to shut down the connection:
+
+```elixir
+test "request with unexpected outage", %{bypass: bypass} do
+  Bypass.down(bypass)
+
+  assert {:error, :econnrefused} = HealthCheck.ping("http://localhost:#{bypass.port}")
+end
+```
+
+If we run our new tests we'll see everything passes as expected!
+With our `HealthCheck` module tested we can move on to testing it together with our GenServer based-scheduler.
+
+## Multiple external hosts
+
+For our project we'll keep the scheduler barebones and rely on `Process.send_after/3` to power our reoccuring checks, for more on the `Process` module take a look at the [documentation](https://hexdocs.pm/elixir/Process.html).
+Our scheduler requires three options: the collection of sites, the interval of our checks, and the module the implements `ping/1`.
+By passing in our module we further decouple our functionality and our GenServer, enabling us to better test each in isolation:
+
+```elixir
+defmodule Clinic.Scheduler do
+  use GenServer
+
+  require Logger
+
+  alias Clinic.HealthCheck
+
+  def init(opts) do
+    sites = Keyword.fetch!(opts, :sites)
+    interval = Keyword.fetch!(opts, :interval)
+
+    Process.send_after(self(), :check, interval)
+
+    {:ok, sites}
+  end
+end
+```
+
+Now we need to define the `handle_info/2` function for the `:check` message sent `send_after/2`.
+To keep things simple we'll pass our sites to `HealthCheck.ping/1` and log our results to either `Logger.info` or in the case of errors `Logger.error`.
+We'll setup our code in a way that will enable us to improve the reporting capabilities at a later time:
+
+```elixir
+def handle_info(:check, sites) do
+  sites
+  |> HealthCheck.ping()
+  |> Enum.each(&report/1)
+end
+
+defp report({:ok, body}), do: Logger.info(body)
+defp report({:error, reason}) do
+  reason
+  |> to_string()
+  |> Logger.error()
+end
+```
+
+As discussed we pass our sites to `HealthCheck.ping/1` then iterate the results with `Enum.each/2` applying our `report/1` function against each.
+With these functions in place our scheduler is done and we can focus on testing it.
+
+We won't focus too much on unit testing the schedulers since that won't require Bypass, so we can skip to the final code:
+
+```elixir
+defmodule Clinic.SchedulerTest do
+  use ExUnit.Case
+
+  import ExUnit.CaptureLog
+
+  alias Clinic.Scheduler
+
+  defmodule TestCheck do
+    def ping(_sites), do: [{:ok, "pong"}, {:error, "HTTP Status 404"}]
+  end
+
+  test "health checks are run and results logged" do
+    opts = [health_checks: TestCheck, interval: 1, sites: ["http://example.com", "http://example.org"]]
+
+    output =
+      capture_log(fn ->
+        {:ok, _pid} = GenServer.start_link(Scheduler, opts)
+        :timer.sleep(10)
+      end)
+
+    assert output =~ "pong"
+    assert output =~ "HTTP Status 404"
+  end
+end
+```
+
```

Comment:
This test does not work for me. Also, running the code, leads me to believe you need to fetch the `health_checks` in `Scheduler.init` and then pass it in `Scheduler.handle_info` as well (so that the test will run with TestCheck and code will run with HealthCheck) and then maybe use `apply` to use the actual function. I might be missing here, but will test it out to be sure.

Reply:
@gemantzu specifically what about the test does not work?  Does it fail?  Does it throw an error?

The module to apply ping to is provided as a option to the GenServer, see the above `health_check` key in the Keyword List.

Reply:
I compared it with your clinic repo, you are missing a line in your `Scheduler.init/1` right after the interval variable fetch.

`health_check = Keyword.get(opts, :health_checks, HealthCheck)`

That's why I was asking how the Scheduler know which module to apply ping to :)

Reply:
Fixed @gemantzu 👍